Pkgconfig: followup changes from PR #373

bluca opened the pull request:
As requested by @jgunthorpe in #373 here's a few more changes: added mlx4/5 pkg-config files, enable IN_PLACE workflow with PKG_CONFIG_FILES, move workaround for CMake + pkg-config quirk into helper function, revert dependency on libibverbs-dev to nl3-dev.

Here's how an IN_PLACE=1 build looks like:

```
$ find build/lib/pkgconfig/
build/lib/pkgconfig/
build/lib/pkgconfig/librdmacm.pc
build/lib/pkgconfig/libmlx5.pc
build/lib/pkgconfig/libibverbs.pc
build/lib/pkgconfig/libmlx4.pc
build/lib/pkgconfig/libibumad.pc
```

```
$ cat build/lib/pkgconfig/libmlx5.pc
prefix=/home/bluca/git/rdma-core/build
exec_prefix=${prefix}
libdir=${prefix}/lib
includedir=${prefix}/include

Name: libmlx5
Description: RDMA Core Userspace Library
Version: 20.0
Libs: -L${libdir} -lmlx5
Libs.private: -lpthread
Cflags: -I${includedir} 
Requires: libibverbs
```

rleon commented:
+1

rleon commented:
@jgunthorpe are you ok with this PR? We need to release rdma-core.

jgunthorpe commented on buildlib/rdma_functions.cmake:
```diff
@@ -269,9 +269,16 @@ endfunction()
+  # With IN_PLACE=1 the install step is not ran, so copy the files manually
+  if (IN_PLACE)
+    configure_file(${CMAKE_CURRENT_BINARY_DIR}/lib${PC_LIB_NAME}.pc ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/pkgconfig/lib${PC_LIB_NAME}.pc COPYONLY)
+  endif()
```
There is no real need to have both .pc files.. Just always emit them to build/lib/pkgconfig

If in_place is set then they are valid for use without installation, otherwise they are the to-be-installed versions..

jgunthorpe replied:
Also the IN_PLACE .pc files should have -Wl,-rpath, I think ?

bluca replied:
> There is no real need to have both .pc files.. Just always emit them to build/lib/pkgconfig
>
> If in_place is set then they are valid for use without installation, otherwise they are the to-be-installed versions..

Ok, done.

> Also the IN_PLACE .pc files should have -Wl,-rpath, I think ?

Good point, yes it should:

```
prefix=/home/bluca/git/rdma-core/build
exec_prefix=${prefix}
libdir=${prefix}/lib
includedir=${prefix}/include

Name: librdmacm
Description: RDMA Core Userspace Library
URL: https://github.com/linux-rdma
Version: 1.1.20.0
Libs: -L${libdir} -lrdmacm -Wl,-rpath,/home/bluca/git/rdma-core/build/lib
Libs.private: 
Cflags: -I${includedir} 
Requires:
```
```
$ gcc /tmp/test.c `PKG_CONFIG_PATH=$PWD/lib/pkgconfig pkg-config --libs --cflags libibverbs`
$ objdump -x a.out | grep RUN
  RUNPATH              /home/bluca/git/rdma-core/build/lib
```

jgunthorpe commented:
The .pc files still need some fixing, the version number and -I thing are particularly important. We can't make a release with wrong .pc metadata, so let us wait on that..

bluca commented:
> The version: field in all the pkgconfigs is wrong, it should be the value of get_target_properties(VERSION) for the library

Ok, but it seems to be working though? Isn't 20.0 the right version?

> The use of libs.private for libnl is also not right, it should be requires.private since libnl is also using pkgconfig.

Requires.private is broken in the sense that it must always be satisfied, even when not using --static. Given you don't want a dependency on libmnl-dev, it would make it unusable as it would always fail, given mnl does not get pulled in.

> The use of
>
> Requires: libibverbs
>
> Everywhere is wrong too, those are all requires.private. AFAIK we have no library with a public requires in rdma-core.

ldd shows the other libraries are linked to it though?

jgunthorpe commented:
>  Ok, but it seems to be working though? Isn't 20.0 the right version?

20 is the version of the overall package, but each library has an individual semantic version (X.Y.20), so if someone wants to use pkgconfig to express dependencies they should use the semantic version, not the overall version.

> Requires.private is broken in the sense that it must always be satisfied, even when not using --static. Given you don't want a dependency on libmnl-dev, it would make it unusable as it would always fail, given mnl does not get pulled in.

That sounds like a defect in whatever is tool processing the pkg-config output. Requires.private is clearly defined as only for static linking, requiring it for dynamic sounds like a tooling bug. I'd be interested if you find any info on this out there..

It is not optional, as if we do statically link, we must have all the pkgconfig data for static linking (eg its *.private stuff) imported from libnl properly.

And no, we should not add a packaging dependencies to work around bugs in other package's tooling, AFAIK.

If this is a trouble to figure out now, we should remove this nl stuff and make a dynamic-link-only pkgconfig, and worry about static linking as part of the static linking support I plan to do this month...

> ldd shows the other libraries are linked to it though?

They are *private*, so the application should never link to them directly when dynamically linking. Doing this is called 'over linking' and it is harmful. The pkg-config stuff is setup with the public/private specifically to try and avoid over-linking when working dynamically.

Apps should only dynamically link to libraries whose symbols they actually consume directly.

The only time an app needs to link explicitly to libibverbs when using a companion library is if the app is statically linking.

bluca commented:
> That sounds like a defect in whatever is tool processing the pkg-config output. Requires.private is clearly defined as only for static linking, requiring it for dynamic sounds like a tooling bug. I'd be interested if you find any info on this out there..

It's a known issue of pkg-config itself, not of  an external tool. There's a bug open on their bugzilla closed as "working as intended" iirc, I'm on the phone right now as I'm traveling, I can look it up on Friday morning.

But it's dead easy to reproduce: add a non existing requires.private and then try to call it without --static and watch it fail due to missing dependency

rleon commented:
FYI,
@jgunthorpe, @bluca I will release tomorrow and will be required to revert previous PR.

dledford commented:
@rleon there is no high priority pressure to release tomorrow.  There is time to sort out these issues, reverting and releasing solely for the sake of releasing is silly.

rleon commented:
IMHO releasing on schedule is a valuable goal.

On Wed, Sep 5, 2018, 16:55 Doug Ledford <notifications@github.com> wrote:

> @rleon <https://github.com/rleon> there is no high priority pressure to
> release tomorrow. There is time to sort out these issues, reverting and
> releasing solely for the sake of releasing is silly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/linux-rdma/rdma-core/pull/374#issuecomment-418739038>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAbhEEB7te5vukjFyxV-IVYW-cjV5Gblks5uX9eogaJpZM4WWnf7>
> .
>

dledford commented:
I would disagree.  If we had some important item relying on releasing on schedule, then maybe so.  But to my knowledge we don't, and just like Linus will run an rc8 when needed, reverting this just to be on time instead of fixing it properly doesn't make sense to me.  Just fix this up and then release.

jgunthorpe commented:
> It's a known issue of pkg-config itself, not of an external tool. There's a bug open on their bugzilla closed as "working as intended" iirc, I'm on the phone right now as I'm traveling, I can look it up on Friday morning.

Well, lets drop libnl completely in pkgconfig and have a dynamic-only pkgconfig. When we add the static library we can decide what do do with the Requires.private..

Honestly, I don't know why people want to use pkgconfig, I thought it was basically discouraged for the same reason .la files turned out to be a bad idea..

bluca commented:
I can send an update in about 24 hours, sorry but I'm traveling without my laptop at the moment so can't right now :-(

rleon commented:
It is ok, Doug was in favor of fixing your previous PR prior to release
and not reverting it. I'm fine with it as long as you remember that
rdma-core release depends on you :)

Thanks

On Wed, Sep 05, 2018 at 04:51:25PM -0700, Luca Boccassi wrote:
> I can send an update in about 24 hours, sorry but I'm traveling without my laptop at the moment so can't right now :-(
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/linux-rdma/rdma-core/pull/374#issuecomment-418917640

bluca commented:
> 20 is the version of the overall package, but each library has an individual semantic version (X.Y.20), so if someone wants to use pkgconfig to express dependencies they should use the semantic version, not the overall version.

Ok, done.

> They are private, so the application should never link to them directly when dynamically linking. Doing this is called 'over linking' and it is harmful. The pkg-config stuff is setup with the public/private specifically to try and avoid over-linking when working dynamically.

Ah yes, of course you are right, removed.

> Well, lets drop libnl completely in pkgconfig and have a dynamic-only pkgconfig. When we add the static library we can decide what do do with the Requires.private..

Ok, dropped. I only added in the first place because I mistakenly thought static builds were already supported.
When you add them, if you decide to use Requires.private, then you'll need to have the dependency as I originally added them (libfoo-dev -> libnl-dev). Otherwise, I'm afraid the flags have to be handled manually as I've done, but that means having some flags (like the -I) that might not be meant for static builds. Unless you want to have further "special cases" dependency by dependency - but they might change with the versions, so it might get messy. Perhaps the Requires.private plus the dependency is the best way in the end, I don't know.

> Honestly, I don't know why people want to use pkgconfig, I thought it was basically discouraged for the same reason .la files turned out to be a bad idea..

Well, because all build systems understand it, and it's very easy to use for dependent projects to get build flags in a standardised and common way. My use case is for DPDK: we are in the process of switching the build system to Meson, and having all dependencies available via pkg-config greatly simplifies things for us.

bluca commented:
@jgunthorpe I think I've applied all the requested changes, please let me know if there's something else I can do.

jgunthorpe commented:
> Well, because all build systems understand it, and it's very easy to use for dependent projects to get build flags in a standardised and common way. My use case is for DPDK: we are in the process of switching the build system to Meson, and having all dependencies available via pkg-config greatly simplifies things for us.

Good libraries don't export flags beyond -lfoo to their users for the dynamic linking case :(

jgunthorpe commented on buildlib/rdma_functions.cmake:
```diff
@@ -269,9 +269,19 @@ endfunction()
+
+  # With IN_PLACE=1 the install step is not ran, so generate the file in the build dir
+  if (IN_PLACE)
+    set(PC_LIB "-Wl,-rpath,${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}")
```
This should just be 

```
-Wl,-rpath,${libdir}
```

Right?

jgunthorpe replied:
And it would be easier to understand what it is as PC_RPATH

jgunthorpe commented on buildlib/template.pc.in:
```diff
@@ -5,8 +5,9 @@ includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@
 Description: RDMA Core Userspace Library
-Version: @PACKAGE_VERSION@
-Libs: -L${libdir} -l@PC_LIB_NAME@
+URL: https://github.com/linux-rdma
```
This should be https://github.com/linux-rdma/rdma-core

jgunthorpe commented on buildlib/rdma_functions.cmake:
```diff
@@ -269,9 +269,19 @@ endfunction()
+  else()
+    configure_file(${CMAKE_SOURCE_DIR}/buildlib/template.pc.in ${CMAKE_CURRENT_BINARY_DIR}/lib${PC_LIB_NAME}.pc @ONLY)
+  endif()
   install(FILES ${CMAKE_CURRENT_BINARY_DIR}/lib${PC_LIB_NAME}.pc DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
```
The install needs to be in the if leg

jgunthorpe commented on librdmacm/CMakeLists.txt:
```diff
@@ -42,4 +42,4 @@ install(TARGETS rspreload DESTINATION "${CMAKE_INSTALL_LIBDIR}/rsocket/")
 rdma_install_symlink("librspreload.so" "${CMAKE_INSTALL_LIBDIR}/rsocket/librspreload.so.1.0.0")
 
-rdma_pkg_config("rdmacm" "" "" "libibverbs")
+rdma_pkg_config("rdmacm" "" "" "")
```
Instead of removing these completely they should be a Requires.private: libibverbs ?

jgunthorpe commented:
I fixed the above and reworked the commit messages a bit.. here is the diff I pushed:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6906deb565db1a..c06f68b57f884d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -371,8 +371,6 @@ if (ENABLE_RESOLVE_NEIGH)
     set(NL_KIND 3)
     set(NL_INCLUDE_DIRS ${NL3_INCLUDE_DIRS})
     set(NL_LIBRARIES ${NL3_LIBRARIES})
-    set(NL_CFLAGS ${NL3_CFLAGS})
-    set(NL_LDFLAGS ${NL3_LDFLAGS})
   else()
     # FIXME: I don't know why we have this fallback, all supported distros
     # have libnl3
@@ -381,8 +379,6 @@ if (ENABLE_RESOLVE_NEIGH)
       set(NL_KIND 1)
       set(NL_INCLUDE_DIRS ${NL1_INCLUDE_DIRS})
       set(NL_LIBRARIES ${NL1_LIBRARIES})
-      set(NL_LDFLAGS ${NL1_LDFLAGS})
-      set(NL_CFLAGS ${NL1_CFLAGS})
     else()
       message(FATAL_ERROR "Cannot find libnl-3.0 or libnl-1")
     endif()
diff --git a/buildlib/rdma_functions.cmake b/buildlib/rdma_functions.cmake
index ba60629b1b4f35..f7c83d106d2aa5 100644
--- a/buildlib/rdma_functions.cmake
+++ b/buildlib/rdma_functions.cmake
@@ -267,21 +267,18 @@ function(rdma_finalize_libs)
 endfunction()
 
 # Generate a pkg-config file
-function(rdma_pkg_config PC_LIB_NAME PC_LIB_PRIVATE PC_INCLUDE_PRIVATE PC_REQUIRES)
+function(rdma_pkg_config PC_LIB_NAME PC_REQUIRES_PRIVATE PC_LIB_PRIVATE)
   set(PC_LIB_NAME "${PC_LIB_NAME}")
-  # CMake separates ldflags with a semicolon, cannot be used in a command
-  string(REPLACE ";" " " PC_LIB_PRIVATE "${PC_LIB_PRIVATE}")
   set(PC_LIB_PRIVATE "${PC_LIB_PRIVATE}")
-  set(PC_INCLUDE_PRIVATE "${PC_INCLUDE_PRIVATE}")
-  set(PC_REQUIRES "${PC_REQUIRES}")
+  set(PC_REQUIRES_PRIVATE "${PC_REQUIRES_PRIVATE}")
   get_target_property(PC_VERSION ${PC_LIB_NAME} VERSION)
 
   # With IN_PLACE=1 the install step is not ran, so generate the file in the build dir
   if (IN_PLACE)
-    set(PC_LIB "-Wl,-rpath,${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}")
+    set(PC_RPATH "-Wl,-rpath,\${libdir}")
     configure_file(${CMAKE_SOURCE_DIR}/buildlib/template.pc.in ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/pkgconfig/lib${PC_LIB_NAME}.pc @ONLY)
   else()
     configure_file(${CMAKE_SOURCE_DIR}/buildlib/template.pc.in ${CMAKE_CURRENT_BINARY_DIR}/lib${PC_LIB_NAME}.pc @ONLY)
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/lib${PC_LIB_NAME}.pc DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
   endif()
-  install(FILES ${CMAKE_CURRENT_BINARY_DIR}/lib${PC_LIB_NAME}.pc DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
 endfunction()
diff --git a/buildlib/template.pc.in b/buildlib/template.pc.in
index 291ccbf22a8994..99a798464abd03 100644
--- a/buildlib/template.pc.in
+++ b/buildlib/template.pc.in
@@ -5,9 +5,9 @@ includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@
 
 Name: lib@PC_LIB_NAME@
 Description: RDMA Core Userspace Library
-URL: https://github.com/linux-rdma
+URL: https://github.com/linux-rdma/rdma-core
 Version: @PC_VERSION@
-Libs: -L${libdir} -l@PC_LIB_NAME@ @PC_LIB@
 Libs.private: @PC_LIB_PRIVATE@
-Cflags: -I${includedir} @PC_INCLUDE_PRIVATE@
-Requires: @PC_REQUIRES@
+Libs: -L${libdir} -l@PC_LIB_NAME@ @PC_RPATH@
+Requires.private: @PC_REQUIRES_PRIVATE@
+Cflags: -I${includedir}
diff --git a/libibumad/CMakeLists.txt b/libibumad/CMakeLists.txt
index 0a0a4abfd08ad0..1f600a0228a234 100644
--- a/libibumad/CMakeLists.txt
+++ b/libibumad/CMakeLists.txt
@@ -16,4 +16,4 @@ rdma_library(ibumad libibumad.map
   umad_str.c
   )
 
-rdma_pkg_config("ibumad" "" "" "")
+rdma_pkg_config("ibumad" "" "")
diff --git a/libibverbs/CMakeLists.txt b/libibverbs/CMakeLists.txt
index 42ba13b2fb8973..738e5e15aaf00e 100644
--- a/libibverbs/CMakeLists.txt
+++ b/libibverbs/CMakeLists.txt
@@ -53,4 +53,4 @@ target_link_libraries(ibverbs LINK_PRIVATE
   kern-abi
   )
 
-rdma_pkg_config("ibverbs" "${CMAKE_THREAD_LIBS_INIT}" "" "")
+rdma_pkg_config("ibverbs" "" "${CMAKE_THREAD_LIBS_INIT}")
diff --git a/librdmacm/CMakeLists.txt b/librdmacm/CMakeLists.txt
index b3e27381273988..f71e4076f53953 100644
--- a/librdmacm/CMakeLists.txt
+++ b/librdmacm/CMakeLists.txt
@@ -42,4 +42,4 @@ install(TARGETS rspreload DESTINATION "${CMAKE_INSTALL_LIBDIR}/rsocket/")
 rdma_install_symlink("librspreload.so" "${CMAKE_INSTALL_LIBDIR}/rsocket/librspreload.so.1")
 rdma_install_symlink("librspreload.so" "${CMAKE_INSTALL_LIBDIR}/rsocket/librspreload.so.1.0.0")
 
-rdma_pkg_config("rdmacm" "" "" "")
+rdma_pkg_config("rdmacm" "libibverbs" "${CMAKE_THREAD_LIBS_INIT}")
diff --git a/providers/mlx4/CMakeLists.txt b/providers/mlx4/CMakeLists.txt
index a10fe083297074..ad849f105231fe 100644
--- a/providers/mlx4/CMakeLists.txt
+++ b/providers/mlx4/CMakeLists.txt
@@ -15,4 +15,4 @@ publish_headers(infiniband
 
 install(FILES "mlx4.conf" DESTINATION "${CMAKE_INSTALL_SYSCONFDIR}/modprobe.d/")
 
-rdma_pkg_config("mlx4" "${CMAKE_THREAD_LIBS_INIT}" "" "")
+rdma_pkg_config("mlx4" "libibverbs" "${CMAKE_THREAD_LIBS_INIT}")
diff --git a/providers/mlx5/CMakeLists.txt b/providers/mlx5/CMakeLists.txt
index d77c8f5e5aad53..4a0f88bcde28a9 100644
--- a/providers/mlx5/CMakeLists.txt
+++ b/providers/mlx5/CMakeLists.txt
@@ -27,4 +27,4 @@ publish_headers(infiniband
   mlx5dv.h
 )
 
-rdma_pkg_config("mlx5" "${CMAKE_THREAD_LIBS_INIT}" "" "")
+rdma_pkg_config("mlx5" "libibverbs" "${CMAKE_THREAD_LIBS_INIT}")
```

bluca commented:
Looks good to me, thanks
